keps/sig-apps/2255-pod-cost/README.md

```diff
@@ -0,0 +1,266 @@
+
+
+### Non-Goals
+
+Guaranteed (in contrast to the recommendation stated in Goals) deletion of a selected replica.
+
+## Proposal
+
+The application can set the `controller.kubernetes.io/pod-cost` annotation to a pod through the Kubernetes API. When a downscale event happens, the pod with the lower priority value of the previously set annotation will be deleted first. If one pod of the Deployment has no priority annotation set, it will be treated as the lowest priority.
+
+If all pods have the same priority, there is no difference in the normal pod delete decision behaviour. The same applies if the delete priority annotation is not used at all.
+
+The delete priority annotation can be changed during operation, for example, if workload changes or a new master gets elected.
+
+### User Stories (optional)
+
+
+#### Story 1
+
+In an application environment with stateful worker (user-)sessions, it is essential to keep the user sessions alive as good as possible. In case of a scale-down event, the application has to tell the scheduler, which delete decision would have the lowest impact on existing sessions.
+
+#### Story 2
+
+An application consists of identical server processes, but one of the replicas will be the master, which should be kept as long as possible. All other replicas can be treated as cattle workload. Then the master can set the priority annotation with a high priority value as soon as it has finished its startup process. The other replicas can remain either without any priority set, or e.g. with all the same, lower priority. This ensures, that the master replica of this deployment will be protected in a downscale situation.
+
+
+### Risks and Mitigations
+
+On previous Kubernetes ReplicaSet controller versions that don't implement the delete priority annotation feature, the same application might make false assumptions about the protection of a master instance or workers with open (user-)sessions on it. As the delete priority annotation would be only a suggestion to the ReplicaSet controller, the application developer should, however, handle the case of a failed master instance or broken user sessions. The feature is just an improvement, not a guarantee, as there might happen timing issues between setting the annotation and the next controller scale-down event.
+
+## Design Details
+
+
+### Test Plan
+
+* Units test in kube-controller-manager package to test a variety of scenarios.
+* New E2E Tests to validate that replicas get deleted as expected e.g:
+ * Replicas with lower delete priority before replicas with higher delete priority
+ * Replicas with no delete priority annotation set before replicas with low priority
+
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+* Implemented feedback from alpha testers
+* Thorough E2E and unit testing in place
+
+#### Beta -> GA Graduation
+* Significant number of end-users are using the feature
```

Comment:
I don't think we need "significant" number :)

Reply:
What else should we put in there?

Reply:
I think this is fine for now.